File: sdks/go/cmd/elydora/plugins/copilot_contract.go
package plugins

import (
	"bytes"
	"encoding/json"
	"fmt"
	"path/filepath"
	"sort"
	"strings"
)

const (
	copilotAgentKey      = "copilot"
	copilotGuardScript   = "guard.js"
	copilotAuditScript   = "hook.js"
	copilotConfigFile    = "elydora-audit.json"
	copilotHookTimeout   = float64(10)
	copilotLegacyTimeout = float64(5)
)

var copilotManagedEvents = []struct {
	event  string
	script string
}{
	{"preToolUse", copilotGuardScript},
	{"postToolUse", copilotAuditScript},
	{"postToolUseFailure", copilotAuditScript},
}

type copilotHooks map[string][]map[string]any

type copilotDocument struct {
	exists        bool
	filePath      string
	root          map[string]any
	hooks         copilotHooks
	raw           []byte
	snapshot      *managedFileSnapshot
	hooksDisabled bool
}

type copilotSourcePrecondition struct {
	filePath string
	label    string
	snapshot *managedFileSnapshot
}

type copilotSources struct {
	user                  *copilotDocument
	legacy                *copilotDocument
	disabledBy            string
	settingsPreconditions []copilotSourcePrecondition
}

type copilotRenderedDocument struct {
	document *copilotDocument
	changed  bool
	next     []byte
	remove   bool
}

type copilotRuntimeContract = managedRuntimeContract

type copilotManagedEntry = managedScriptReference

func copilotManagedEntryForHandler(
	handler map[string]any,
	scriptName string,
) (*copilotManagedEntry, error) {
	scriptPath, managed := copilotManagedScriptPath(handler)
	if !managed {
		return nil, nil
	}
	return resolveManagedScript(scriptPath, scriptName)
}

func cloneCopilotHooks(source copilotHooks) copilotHooks {
	clone := make(copilotHooks, len(source))
	for event, handlers := range source {
		copied := make([]map[string]any, len(handlers))
		copy(copied, handlers)
		clone[event] = copied
	}
	return clone
}

func buildCopilotHandler(nodePath, scriptPath string) map[string]any {
	return map[string]any{
		"type":       "command",
		"bash":       posixSource(nodePath, scriptPath),
		"powershell": powerShellSource(nodePath, scriptPath),
		"timeoutSec": copilotHookTimeout,
	}
}

func parseCopilotLegacyCommand(command string) (string, bool) {
	command = strings.TrimSpace(command)
	separator := strings.IndexAny(command, " \t")
	if separator < 0 {
		return "", false
	}
	executable := command[:separator]
	if !strings.EqualFold(executable, "node") && !strings.EqualFold(executable, "node.exe") {
		return "", false
	}
	script := strings.TrimSpace(command[separator:])
	if len(script) >= 2 && ((script[0] == '"' && script[len(script)-1] == '"') ||
		(script[0] == '\'' && script[len(script)-1] == '\'')) {
		script = script[1 : len(script)-1]
	}
	return script, script != ""
}

func copilotManagedScriptPath(handler map[string]any) (string, bool) {
	if len(handler) != 4 || handler["type"] != "command" {
		return "", false
	}
	bash, bashOK := handler["bash"].(string)
	powershell, powershellOK := handler["powershell"].(string)
	if !bashOK || !powershellOK {
		return "", false
	}
	if handler["timeoutSec"] == copilotHookTimeout {
		bashNode, bashScript, bashParsed := parsePOSIXCommand(bash)
		psNode, psScript, psParsed := parsePowerShellSource(powershell)
		if bashParsed && psParsed && filepath.IsAbs(bashNode) && filepath.IsAbs(bashScript) &&
			isNodeExecutable(bashNode) && sameManagedPath(bashNode, psNode) &&
			sameManagedPath(bashScript, psScript) {
			return bashScript, true
		}
	}
	if handler["timeoutSec"] != copilotLegacyTimeout || bash != powershell {
		return "", false
	}
	return parseCopilotLegacyCommand(bash)
}

func removeManagedCopilotHooks(
	hooks copilotHooks,
	agentID string,
) (copilotHooks, error) {
	result := cloneCopilotHooks(hooks)
	for _, contract := range copilotManagedEvents {
		kept := make([]map[string]any, 0, len(result[contract.event]))
		for _, handler := range result[contract.event] {
			entry, err := copilotManagedEntryForHandler(handler, contract.script)
			if err != nil {
				return nil, err
			}
			remove := entry != nil &&
				(agentID == "" || sameManagedAgentID(entry.agentID, agentID))
			if !remove {
				kept = append(kept, handler)
			}
		}
		if len(kept) == 0 {
			delete(result, contract.event)
		} else {
			result[contract.event] = kept
		}
	}
	return result, nil
}

func parseCopilotDocument(
	filePath string,
	snapshot *managedFileSnapshot,
	label string,
) (*copilotDocument, error) {
	if snapshot == nil {
		return nil, fmt.Errorf("%s snapshot is required: %s", label, filePath)
	}
	documentLabel := fmt.Sprintf("%s at %s", label, filePath)
	root, err := decodeStrictJSONObject(snapshot.contents, documentLabel)
	if err != nil {
		return nil, err
	}
	if root["version"] != float64(1) {
		return nil, fmt.Errorf("%s must declare version 1", documentLabel)
	}
	disabled := false
	if value, exists := root["disableAllHooks"]; exists {
		var valid bool
		disabled, valid = value.(bool)
		if !valid {
			return nil, fmt.Errorf(`%s field "disableAllHooks" must be a boolean`, documentLabel)
		}
	}
	hooks := copilotHooks{}
	if value, exists := root["hooks"]; exists {
		hooks, err = validateCopilotHooks(value, documentLabel)
		if err != nil {
			return nil, err
		}
	}
	return &copilotDocument{
		exists: true, filePath: filePath, root: root, hooks: hooks,
		raw: append([]byte(nil), snapshot.contents...), snapshot: snapshot,
		hooksDisabled: disabled,
	}, nil
}

func createCopilotDocument(filePath string) *copilotDocument {
	return &copilotDocument{
		filePath: filePath,
		root:     map[string]any{},
		hooks:    copilotHooks{},
	}
}

func emptyOwnedCopilotDocument(root map[string]any, hooks copilotHooks) bool {
	if len(hooks) != 0 {
		return false
	}
	for key := range root {
		if key != "version" && key != "hooks" {
			return false
		}
	}
	return true
}

func renderCopilotDocument(
	document *copilotDocument,
	hooks copilotHooks,
) (*copilotRenderedDocument, error) {
	if !document.exists && len(hooks) == 0 {
		return &copilotRenderedDocument{document: document}, nil
	}
	if document.exists && emptyOwnedCopilotDocument(document.root, hooks) {
		return &copilotRenderedDocument{
			document: document, changed: true, remove: true,
		}, nil
	}
	root := cloneJSONObject(document.root)
	root["version"] = float64(1)
	if len(hooks) == 0 {
		delete(root, "hooks")
	} else {
		root["hooks"] = hooks
	}
	next, err := json.MarshalIndent(root, "", "  ")
	if err != nil {
		return nil, fmt.Errorf("encode GitHub Copilot hooks: %w", err)
	}
	next = append(next, '\n')
	if _, err := parseCopilotDocument(
		document.filePath,
		&managedFileSnapshot{contents: next},
		"GitHub Copilot rendered hooks",
	); err != nil {
		return nil, fmt.Errorf("validate rendered GitHub Copilot hooks: %w", err)
	}
	return &copilotRenderedDocument{
		document: document,
		changed:  !document.exists || !bytes.Equal(next, document.raw),
		next:     next,
	}, nil
}

func copilotManagedEntries(
	handlers []map[string]any,
	scriptName string,
) (map[string][]copilotManagedEntry, error) {
	result := map[string][]copilotManagedEntry{}
	for _, handler := range handlers {
		entry, err := copilotManagedEntryForHandler(handler, scriptName)
		if err != nil {
			return nil, err
		}
		if entry != nil {
			key := managedReferenceKey(entry.agentID)
			result[key] = append(result[key], *entry)
		}
	}
	return result, nil
}

func copilotRuntimeContracts(hooks copilotHooks) ([]copilotRuntimeContract, error) {
	guards, err := copilotManagedEntries(hooks["preToolUse"], copilotGuardScript)
	if err != nil {
		return nil, err
	}
	successes, err := copilotManagedEntries(hooks["postToolUse"], copilotAuditScript)
	if err != nil {
		return nil, err
	}
	failures, err := copilotManagedEntries(
		hooks["postToolUseFailure"], copilotAuditScript,
	)
	if err != nil {
		return nil, err
	}
	keys := make([]string, 0, len(guards))
	for key := range guards {
		keys = append(keys, key)
	}
	sort.Strings(keys)
	contracts := make([]copilotRuntimeContract, 0, len(keys))
	for _, key := range keys {
		guard := guards[key]
		success := successes[key]
		failure := failures[key]
		if len(guard) != 1 || len(success) != 1 || len(failure) != 1 ||
			!sameManagedPath(success[0].scriptPath, failure[0].scriptPath) {
			continue
		}
		contracts = append(contracts, copilotRuntimeContract{
			agentID: guard[0].agentID, guardPath: guard[0].scriptPath,
			auditPath: success[0].scriptPath,
		})
	}
	return contracts, nil
}

File: sdks/node/src/plugins/copilot-contract.ts
import os from 'node:os';
import path from 'node:path';
import { sameAgentId, samePath } from './common.js';
import type { FileSnapshot } from './managed-files.js';
import {
  isNodeExecutable,
  parsePosixCommand,
  parsePowerShellSource,
  posixSource,
  powerShellSource,
} from './shell-command.js';
import { validateHooks, type CopilotHooks } from './copilot-schema.js';
import { parseStrictJsonObject, type JsonObject } from './strict-json.js';

export const AGENT_KEY = 'copilot';
export const GUARD_SCRIPT = 'guard.js';
export const AUDIT_SCRIPT = 'hook.js';
export const HOOK_TIMEOUT_SECONDS = 10;
export const CONFIG_FILE = 'elydora-audit.json';

const MANAGED_EVENTS = [
  ['preToolUse', GUARD_SCRIPT],
  ['postToolUse', AUDIT_SCRIPT],
  ['postToolUseFailure', AUDIT_SCRIPT],
] as const;

export interface CopilotDocument {
  readonly exists: boolean;
  readonly filePath: string;
  readonly root: JsonObject;
  readonly hooks: CopilotHooks;
  readonly hooksDisabled: boolean;
  readonly raw?: string;
  readonly snapshot?: FileSnapshot;
}

export interface CopilotSources {
  readonly user: CopilotDocument;
  readonly legacy?: CopilotDocument;
  readonly disabledBy?: string;
  readonly settingsPreconditions: readonly CopilotSourcePrecondition[];
}

export interface CopilotSourcePrecondition {
  readonly filePath: string;
  readonly label: string;
  readonly snapshot?: FileSnapshot;
}

export interface RenderedDocument {
  readonly document: CopilotDocument;
  readonly changed: boolean;
  readonly next?: string;
}

export interface RuntimeContract {
  readonly agentId: string;
  readonly guardPath: string;
  readonly auditPath: string;
}

interface ManagedEntry {
  readonly agentId: string;
  readonly scriptPath: string;
}

function agentKey(agentId: string): string {
  return process.platform === 'win32' ? agentId.toLowerCase() : agentId;
}

export function buildHandler(scriptPath: string): JsonObject {
  return {
    type: 'command',
    bash: posixSource(scriptPath),
    powershell: powerShellSource(scriptPath),
    timeoutSec: HOOK_TIMEOUT_SECONDS,
  };
}

function exactHandlerKeys(handler: JsonObject): boolean {
  return Object.keys(handler).sort().join(',') === 'bash,powershell,timeoutSec,type';
}

function currentManagedScriptPath(handler: JsonObject): string | undefined {
  if (!exactHandlerKeys(handler)
    || handler.type !== 'command'
    || handler.timeoutSec !== HOOK_TIMEOUT_SECONDS) return undefined;
  if (typeof handler.bash !== 'string' || typeof handler.powershell !== 'string') return undefined;
  const bash = parsePosixCommand(handler.bash);
  const powershell = parsePowerShellSource(handler.powershell);
  if (!bash || !powershell
    || !path.isAbsolute(bash[0])
    || !isNodeExecutable(bash[0])
    || !samePath(bash[0], powershell[0])
    || !samePath(bash[1], powershell[1])) return undefined;
  return bash[1];
}

function parseLegacyCommand(command: unknown): string | undefined {
  if (typeof command !== 'string') return undefined;
  return /^node(?:\.exe)?\s+"([^"]+)"$/i.exec(command)?.[1];
}

function legacyManagedScriptPath(handler: JsonObject): string | undefined {
  if (!exactHandlerKeys(handler)
    || handler.type !== 'command'
    || handler.timeoutSec !== 5) return undefined;
  const bash = parseLegacyCommand(handler.bash);
  const powershell = parseLegacyCommand(handler.powershell);
  return bash && powershell && samePath(bash, powershell) ? bash : undefined;
}

function managedEntry(handler: JsonObject, scriptName: string): ManagedEntry | undefined {
  const scriptPath = currentManagedScriptPath(handler) ?? legacyManagedScriptPath(handler);
  if (!scriptPath || path.basename(scriptPath).toLowerCase() !== scriptName.toLowerCase()) {
    return undefined;
  }
  const agentDirectory = path.dirname(scriptPath);
  if (!samePath(path.dirname(agentDirectory), path.join(os.homedir(), '.elydora'))) {
    return undefined;
  }
  const agentId = path.basename(agentDirectory);
  if (!agentId || agentId === '.' || agentId === '..') return undefined;
  return { agentId, scriptPath };
}

export function parseDocument(
  filePath: string,
  snapshot: FileSnapshot,
  label: string,
): CopilotDocument {
  const root = parseStrictJsonObject(snapshot.contents, `${label} at ${filePath}`);
  if (root.version !== 1) throw new Error(`${label} at ${filePath} must declare version 1`);
  if (root.disableAllHooks !== undefined && typeof root.disableAllHooks !== 'boolean') {
    throw new Error(`${label} at ${filePath} field "disableAllHooks" must be a boolean`);
  }
  return {
    exists: true,
    filePath,
    root,
    hooks: validateHooks(root.hooks, `${label} at ${filePath}`),
    hooksDisabled: root.disableAllHooks === true,
    raw: snapshot.contents,
    snapshot,
  };
}

export function createDocument(filePath: string): CopilotDocument {
  return { exists: false, filePath, root: {}, hooks: {}, hooksDisabled: false };
}

export function removeManagedHooks(hooks: CopilotHooks, agentId?: string): CopilotHooks {
  const next: CopilotHooks = { ...hooks };
  for (const [event, scriptName] of MANAGED_EVENTS) {
    const handlers = (next[event] ?? []).filter((handler) => {
      const owned = managedEntry(handler, scriptName);
      return !owned || (agentId !== undefined && !sameAgentId(owned.agentId, agentId));
    });
    if (handlers.length > 0) next[event] = handlers;
    else delete next[event];
  }
  return next;
}

function isEmptyOwnedDocument(root: JsonObject, hooks: CopilotHooks): boolean {
  return Object.keys(hooks).length === 0
    && Object.keys(root).every((key) => key === 'version' || key === 'hooks');
}

export function renderDocument(
  document: CopilotDocument,
  hooks: CopilotHooks,
): RenderedDocument {
  if (!document.exists && Object.keys(hooks).length === 0) {
    return { document, changed: false };
  }
  if (document.exists && isEmptyOwnedDocument(document.root, hooks)) {
    return { document, changed: true };
  }
  const root: JsonObject = { ...document.root, version: 1 };
  if (Object.keys(hooks).length > 0) root.hooks = hooks;
  else delete root.hooks;
  const next = `${JSON.stringify(root, null, 2)}\n`;
  return { document, changed: next !== document.raw, next };
}

function managedEntries(handlers: JsonObject[], scriptName: string): Map<string, ManagedEntry> {
  const entries = new Map<string, ManagedEntry>();
  for (const handler of handlers) {
    const entry = managedEntry(handler, scriptName);
    if (entry) entries.set(agentKey(entry.agentId), entry);
  }
  return entries;
}

export function runtimeContracts(hooks: CopilotHooks): RuntimeContract[] {
  const guards = managedEntries(hooks.preToolUse ?? [], GUARD_SCRIPT);
  const successes = managedEntries(hooks.postToolUse ?? [], AUDIT_SCRIPT);
  const failures = managedEntries(hooks.postToolUseFailure ?? [], AUDIT_SCRIPT);
  const contracts: RuntimeContract[] = [];
  for (const [key, guard] of guards) {
    const success = successes.get(key);
    const failure = failures.get(key);
    if (!success || !failure || !samePath(success.scriptPath, failure.scriptPath)) continue;
    contracts.push({
      agentId: guard.agentId,
      guardPath: guard.scriptPath,
      auditPath: success.scriptPath,
    });
  }
  return contracts;
}
